src/main/java/de/retest/recheck/auth/RetestAuthentication.java

```diff
@@ -0,0 +1,235 @@
+		parseAccessToken( tokenResponse );
+	}
+
+	private void parseAccessToken( final AccessTokenResponse tokenResponse ) throws VerificationException {
+		accessTokenString = tokenResponse.getToken();
+		refreshToken = tokenResponse.getRefreshToken();
+		final String idTokenString = tokenResponse.getIdToken();
+
+		final AdapterTokenVerifier.VerifiedTokens tokens =
+				AdapterTokenVerifier.verifyTokens( accessTokenString, idTokenString, deployment );
+		accessToken = tokens.getAccessToken();
+	}
+
+	public AccessToken getAccessToken() {
+		refreshTokens();
+		return accessToken;
+	}
+
+	public String getAccessTokenString() {
+		refreshTokens();
+		return accessTokenString;
+	}
+
+	public String getRefreshTokenString() {
+		refreshTokens();
+		return refreshToken;
+	}
+
+	private void refreshTokens() {
+		if ( !isTokenValid() ) {
+			try {
+				parseAccessToken( ServerRequest.invokeRefresh( deployment, refreshToken ) );
+			} catch ( final IOException | HttpFailure | VerificationException e ) {
+				log.error( "Error refreshing token(s)", e );
+			}
+		}
+	}
+
+	private boolean isTokenValid() {
+		try {
+			return accessTokenString != null || accessToken != null
+					|| verifyToken( accessTokenString, deployment ).isActive();
+		} catch ( final VerificationException e ) {
+			log.error( "Error verifying token(s)", e );
+		}
+		return false;
+	}
+
+	static KeycloakResult getRequestParameters( final String request ) {
+		final HttpUrl url = HttpUrl.parse( "http://localhost/" ).resolve( request.split( " " )[1] );
+
+		return KeycloakResult.builder() //
+				.code( url.queryParameter( OAuth2Constants.CODE ) ) //
+				.error( url.queryParameter( OAuth2Constants.ERROR ) ) //
+				.errorDescription( url.queryParameter( "error-description" ) ) //
+				.state( url.queryParameter( OAuth2Constants.STATE ) )//
+				.build();
+	}
+
+	private class CallbackListener extends Thread {
+
+		private final ServerSocket server;
+		private KeycloakResult result;
+
+		public CallbackListener() throws IOException {
+			server = new ServerSocket( 0 );
+		}
+
+		@Override
+		public void run() {
+			try ( Socket socket = server.accept() ) {
+				@Cleanup
+				final BufferedReader br = new BufferedReader( new InputStreamReader( socket.getInputStream() ) );
+				final String request = br.readLine();
+
+				result = getRequestParameters( request );
+
+				@Cleanup
+				final OutputStreamWriter out = new OutputStreamWriter( socket.getOutputStream() );
+				@Cleanup
+				final PrintWriter writer = new PrintWriter( out );
+
+				if ( result.getError() == null ) {
+					writer.println( "HTTP/1.1 302 Found" );
+					writer.println( "Location: " + deployment.getTokenUrl().replace( "/token", "/delegated" ) );
+
+				} else {
+					writer.println( "HTTP/1.1 302 Found" );
+					writer.println(
+							"Location: " + deployment.getTokenUrl().replace( "/token", "/delegated?error=true" ) );
+
+				}
+				writer.flush();
```

Comment:
If `@Cleanup` does `close()`, I believe there is no need for `flush()` as `close()` does this automatically.